src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/TagHelperFactsService.cs

```diff
@@ -15,5 +16,73 @@ public abstract class TagHelperFactsService
 
         public abstract IReadOnlyList<TagHelperDescriptor> GetTagHelpersGivenParent(TagHelperDocumentContext documentContext, string parentTag);
+
+        // Internal for testing
+        internal virtual IEnumerable<KeyValuePair<string, string>> StringifyAttributes(SyntaxList<RazorSyntaxNode> attributes)
+        {
+            var stringifiedAttributes = new List<KeyValuePair<string, string>>();
+
+            for (var i = 0; i < attributes.Count; i++)
+            {
+                var attribute = attributes[i];
+                if (attribute is MarkupTagHelperAttributeSyntax tagHelperAttribute)
+                {
+                    var name = tagHelperAttribute.Name.GetContent();
+                    var value = tagHelperAttribute.Value?.GetContent() ?? string.Empty;
+                    stringifiedAttributes.Add(new KeyValuePair<string, string>(name, value));
+                }
+                else if (attribute is MarkupMinimizedTagHelperAttributeSyntax minimizedTagHelperAttribute)
+                {
+                    var name = minimizedTagHelperAttribute.Name.GetContent();
+                    stringifiedAttributes.Add(new KeyValuePair<string, string>(name, string.Empty));
+                }
+                else if (attribute is MarkupAttributeBlockSyntax markupAttribute)
+                {
+                    var name = markupAttribute.Name.GetContent();
+                    var value = markupAttribute.Value?.GetContent() ?? string.Empty;
+                    stringifiedAttributes.Add(new KeyValuePair<string, string>(name, value));
+                }
+                else if (attribute is MarkupMinimizedAttributeBlockSyntax minimizedMarkupAttribute)
+                {
+                    var name = minimizedMarkupAttribute.Name.GetContent();
+                    stringifiedAttributes.Add(new KeyValuePair<string, string>(name, string.Empty));
+                }
+                else if (attribute is MarkupTagHelperDirectiveAttributeSyntax directiveAttribute)
+                {
+                    var name = directiveAttribute.FullName;
+                    var value = directiveAttribute.Value?.GetContent() ?? string.Empty;
+                    stringifiedAttributes.Add(new KeyValuePair<string, string>(name, value));
+                }
+                else if (attribute is MarkupMinimizedTagHelperDirectiveAttributeSyntax minimizedDirectiveAttribute)
+                {
+                    var name = minimizedDirectiveAttribute.FullName;
+                    stringifiedAttributes.Add(new KeyValuePair<string, string>(name, string.Empty));
+                }
+            }
+
+            return stringifiedAttributes;
+        }
+
+        // Internal for testing
+        internal virtual (string ancestorTagName, bool ancestorIsTagHelper) GetNearestAncestorTagInfo(IEnumerable<SyntaxNode> ancestors)
```

Comment:
I like it. Can this just be a private(internal for testing) method where it is used? I don't see a lot of value in this staying in `TagHelperFactsService`. What do you guys think?

Reply:
Ignore me. I just noticed this moved from CompletionService. This is good. Move the implementation to the derived class though.